lib/grpramat.gi

```diff
@@ -249,109 +249,61 @@ InstallMethod( IsFinite,
-        for pnt in orb do
-            for gen in [1..Length(gens)] do
-                img := pnt * gensp[gen];
-                if not img in orb  then
-                    Add( orb, img );
-                    tmp := rep[ Position( orb, pnt ) ] * gens[gen];
-                    # simple test for infinite order
-                    # Order() would be too expensive to do on all elements
-                    if AbsInt( TraceMat( tmp ) ) > dim then
-                        return false;
-                    fi;
-                    Add( rep, tmp );
-                else
-                    sch := rep[ Position( orb, pnt ) ] * gens[gen]
-                           / rep[ Position( orb, img ) ];
-                    if i = dim then
-                        if sch <> One( grp ) then
-                            if sch * sch <> One( grp ) then
-                                return false;
-                            fi;
-                            if ForAny( stb, x -> x * sch <> sch * x ) then
-                                return false;
-                            fi;
-                        fi;
-                    else
-                        # simple test for infinite order
-                        # Order() would be too expensive to do on all elements
-                        if AbsInt( TraceMat( sch ) ) > dim then
-                            return false;
-                        fi;
-                    fi;
-                    AddSet( stb, sch );
+    badPrimes := [ 2 ];
+    n := DimensionOfMatrixGroup( G );
+    FindPrimesInMatDenominators := function( mat )
+        local i, j, d;
+        for i in [1..n] do
+            for j in [1..n] do
+                d := DenominatorRat(mat[i,j]);
+                if d > 1 then
+                    UniteSet(badPrimes, PrimeDivisors(d));
                 fi;
             od;
         od;
-        grp  := GroupByGenerators( stb, One( grp ) );
-        size := size * Length( orb );
+    end;
+    for g in GeneratorsOfGroup( G ) do
+        FindPrimesInMatDenominators(g);
+        FindPrimesInMatDenominators(g^-1);
     od;
 
-    # if we arrive here, the group is finite
-    SetIsFinite( grp, true );
-    SetSize( G, size * Size( grp ) );
-    return true;
+    p := 3;
+    while p in badPrimes do
+        p := NextPrimeInt(p);
+    od;
+
+    # now reduce mod p
+    e := One(GF(p));
+    H := Group( GeneratorsOfGroup( G ) * e );
 
+    # TODO: could speed up things by checking Minkowski bounds here to
+    # immediately reject some G as infinite -- at least for small enough n
+
+    # evaluate relators
+    phi := IsomorphismFpGroupByGenerators(H, GeneratorsOfGroup( H ));
+
+    gens := GeneratorsOfGroup(FreeGroupOfFpGroup(Range(phi)));
+    rels := RelatorsOfFpGroup(Range(phi));
+    if not ForAll(rels, r -> IsOne(MappedWord(r, gens, GeneratorsOfGroup(G)))) then
+        return false;
+    fi;
+
+    # set as a nice monomorphism
+    phi := GroupHomomorphismByFunction(G, H,  x -> x * e);
```

Comment:
This way, `phi` does not have the information how to compute preimages.
In Oscar, we proceed as follows: We have already computed an isomorphism to a f.p. group on the generators of `H`, and these generators correspond to the generators of `G`. Now apply this isomorphism, and evaluate the word in the generators of `G`.